src/solvers/refinement/string_constraint_generator_float.cpp

```diff
@@ -0,0 +1,454 @@
+/// double precision and floating point value that exceed 100000
+/// \param f: expression representing a float
+/// \param ref_type: refined type for strings
+/// \return a new string expression
+string_exprt string_constraint_generatort::add_axioms_for_string_of_float(
+  const exprt &f, const refined_string_typet &ref_type)
+{
+  const floatbv_typet &type=to_floatbv_type(f.type());
+  ieee_float_spect float_spec(type);
+
+  // We will look at the first 5 digits of the fractional part.
+  // shifted is floor(f * 1e5)
+  exprt shifting=constant_float(1e5, float_spec);
+  exprt shifted=round_expr_to_zero(floatbv_mult(f, shifting));
+  // Numbers with greater or equal value to the following, should use
+  // the exponent notation.
+  exprt max_non_exponent_notation=from_integer(100000, shifted.type());
+  // fractional_part is floor(f * 100000) % 100000
+  mod_exprt fractional_part(shifted, max_non_exponent_notation);
+  string_exprt fractional_part_str=
+    add_axioms_for_fractional_part(fractional_part, 6, ref_type);
+
+  // The axiom added to convert to integer should always be satisfiable even
+  // when the preconditions are not satisfied.
+  mod_exprt integer_part(round_expr_to_zero(f), max_non_exponent_notation);
+  // We should not need more than 8 characters to represent the integer
+  // part of the float.
+  string_exprt integer_part_str=add_axioms_from_int(integer_part, 8, ref_type);
+
+  return add_axioms_for_concat(integer_part_str, fractional_part_str);
+}
+
+/// Add axioms for representing the fractional part of a floating point starting
+/// with a dot
+/// \param int_expr: an integer expression
+/// \param max_size: a maximal size for the string, this includes the
+///   potential minus sign and therefore should be greater than 2
+/// \param ref_type: a type for refined strings
+/// \return a string expression
+string_exprt string_constraint_generatort::add_axioms_for_fractional_part(
+  const exprt &int_expr, size_t max_size, const refined_string_typet &ref_type)
+{
+  PRECONDITION(int_expr.type().id()==ID_signedbv);
+  PRECONDITION(max_size>=2);
+  const typet &type=int_expr.type();
+  string_exprt res=fresh_string(ref_type);
+  exprt ten=from_integer(10, type);
+  const typet &char_type=ref_type.get_char_type();
+  const typet &index_type=ref_type.get_index_type();
+  exprt zero_char=constant_char('0', char_type);
+  exprt nine_char=constant_char('9', char_type);
+  exprt max=from_integer(max_size, index_type);
+
+  // We add axioms:
+  // a1 : 2 <= |res| <= max_size
+  // a2 : starts_with_dot && no_trailing_zero && is_number
+  // starts_with_dot: res[0] = '.'
+  // is_number: forall i:[1, max_size[. '0' < res[i] < '9'
+  // no_trailing_zero: forall j:[2, max_size[. !(|res| = j+1 && res[j]='0')
+  // a3 : int_expr = sum_j 10^j (j < |res| ? res[j] - '0' : 0)
+
+  and_exprt a1(res.axiom_for_is_strictly_longer_than(1),
+               res.axiom_for_is_shorter_than(max));
```

Comment:
Possibly in a different PR: "is shorter than" means <, not <=. I see the actual function does implement <= however. Suggest renaming `axiom_for_shorter_than` and `axiom_for_shorter_or_equal` and similarly >.